lib/teleterm/apiserver/apiserver.go

```diff
@@ -92,3 +111,58 @@ type APIServer struct {
 	// grpc is an instance of grpc server
 	grpcServer *grpc.Server
 }
+
+func getGrpcCredentials(cfg Config) (grpc.ServerOption, error) {
+	uri, err := utils.ParseAddr(cfg.HostAddr)
+
+	if err != nil {
+		return nil, trace.BadParameter("invalid host address: %s", cfg.HostAddr)
+	}
+
+	if uri.Network() != "unix" {
+		keyPair, err := generateKeyPair(cfg.CertsDir)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		return grpc.Creds(keyPair), nil
+	}
+
+	return grpc.Creds(nil), nil
+}
+
+func generateKeyPair(certsDir string) (credentials.TransportCredentials, error) {
+	cert, err := utils.GenerateSelfSignedCert([]string{"localhost"})
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to generate a certificate")
+	}
+
+	err = os.WriteFile(filepath.Join(certsDir, "tsh_server.crt"), cert.Cert, 0600)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to save server certificate")
+	}
+
+	certificate, err := tls.X509KeyPair(cert.Cert, cert.PrivateKey)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to parse server certificates")
+	}
+
+	tlsConfig := &tls.Config{
+		GetConfigForClient: func(info *tls.ClientHelloInfo) (*tls.Config, error) {
+			caCert, err := os.ReadFile(filepath.Join(certsDir, "client.crt"))
```

Comment:
Can we add a check for `certsDir`'s permissions here? Other tools like `ssh` do something along those lines as well, they refuse to work if some sensitive directories or files have permissions that are too lax.

Reply:
Looks like checking permissions on Windows is kind of a mess - thankfully with just the certs on disk rather than the keys we only require that other users are not able to write in our directory, which should be the default for `%APPDATA%`.

Reply:
Yes, a non-privileged user needs a password to access other user data.